transaction-decoder-tool/src/test/kotlin/linea/test/BlockEncodingValidator.kt

```diff
@@ -0,0 +1,179 @@
+import kotlin.time.Duration.Companion.milliseconds
+
+// 100MB, much larger than a real blob, but just for testing to allow faster testing by compressing more blocks
+val BLOB_COMPRESSOR_SIZE: UInt = 100u * 1024u * 1024U
+class BlockEncodingValidator(
+  val vertx: Vertx,
+  val compressorVersion: BlobCompressorVersion = BlobCompressorVersion.V1_0_1,
+  val decompressorVersion: BlobDecompressorVersion = BlobDecompressorVersion.V1_1_0,
+  val blobSizeLimitBytes: UInt = BLOB_COMPRESSOR_SIZE,
+  val log: Logger = LogManager.getLogger(BlockEncodingValidator::class.java)
+) : PeriodicPollingService(vertx, pollingIntervalMs = 1.milliseconds.inWholeMilliseconds, log = log) {
+
+  val compressor = GoBackedBlobCompressor.getInstance(compressorVersion, blobSizeLimitBytes)
+  val decompressor = GoNativeBlobDecompressorFactory.getInstance(decompressorVersion)
+  val rlpEncoder = BesuRlpMainnetEncoderAsyncVertxImpl(vertx)
+  val rlpMainnetDecoder = BesuRlpDecoderAsyncVertxImpl.mainnetDecoder(vertx)
+  val rlpBlobDecoder = BesuRlpDecoderAsyncVertxImpl.blobDecoder(vertx)
+  val queueOfBlocksToValidate = ConcurrentLinkedQueue<Block>()
+  var highestValidatedBlockNumber = AtomicReference<ULong>(ULong.MIN_VALUE)
+
+  override fun action(): SafeFuture<*> {
+    return validateCycle()
+  }
+
+  fun validateRlpEncodingDecoding(blocks: List<Block>): SafeFuture<Unit> {
+    val besuBlocks = blocks.map { it.toBesu() }
+    return rlpEncoder.encodeAsync(besuBlocks)
+      .thenCompose { encodedBlocks ->
+        rlpMainnetDecoder.decodeAsync(encodedBlocks)
+      }
+      .thenApply { decodedBlocks ->
+        val unMatchingBlocks = besuBlocks.zip(decodedBlocks).filter { (expected, actual) -> expected != actual }
+        if (unMatchingBlocks.isEmpty()) {
+          log.info(
+            "all blocks encoding/decoding match: blocks={}",
+            CommonDomainFunctions.blockIntervalString(blocks.first().number, blocks.last().number)
+          )
+        } else {
+          unMatchingBlocks.forEach { (expected, actual) ->
+            log.error(
+              "block encoding/decoding mismatch: block={} \nexpected={} \nactual={}",
+              expected.header.number,
+              expected,
+              actual
+            )
+          }
+        }
+      }
+  }
+
+  fun validateCompression(blocks: List<Block>): SafeFuture<Unit> {
+    queueOfBlocksToValidate.addAll(blocks)
+    return SafeFuture.completedFuture(Unit)
+  }
+
+  fun validateCycle(): SafeFuture<Unit> {
+    val blocks = queueOfBlocksToValidate.pull(300)
+    if (blocks.isEmpty()) {
+      return SafeFuture.completedFuture(Unit)
+    }
+    log.info(
+      "compression validation blocks={} started",
+      CommonDomainFunctions.blockIntervalString(blocks.first().number, blocks.last().number)
+    )
+    val besuBlocks = blocks.map { it.toBesu() }
+    val originalBlockInterval = CommonDomainFunctions.blockIntervalString(blocks.first().number, blocks.last().number)
```

Comment:
nit: This could be moved closer to failure message and message itself can be made lazy

Reply:
I will keep it as it's used below as well (ti was meant to be before).